crates/shared/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -86,166 +101,352 @@ impl TryFrom<PoolData> for PoolInfo {
+        let last_event_block = self.events.lock().await.store().last_event_block().await?;
+        let old_checkpoint_block = self
+            .checkpoint
+            .pools_checkpoint
+            .lock()
+            .unwrap()
+            .block_number;
+        let new_checkpoint_block = std::cmp::max(
+            last_event_block.saturating_sub(MAX_REORG_BLOCK_COUNT),
+            old_checkpoint_block,
+        );
+
+        if new_checkpoint_block > old_checkpoint_block {
+            {
+                let block_range =
+                    RangeInclusive::try_new(old_checkpoint_block + 1, new_checkpoint_block)?;
+                let events = self.events.lock().await.store().get_events(block_range);
+                let mut checkpoint = self.checkpoint.pools_checkpoint.lock().unwrap();
+                append_events(&mut checkpoint.pools, events);
+                checkpoint.block_number = new_checkpoint_block;
+            }
+
+            // clear events before new_checkpoint_block
+            self.events
+                .lock()
+                .await
+                .store_mut()
+                .remove_events_older_than_block(new_checkpoint_block);
+        }
+        Ok(())
     }
 }
 
 #[async_trait::async_trait]
-impl Maintaining for UniswapV3PoolFetcher {
-    async fn run_maintenance(&self) -> Result<()> {
-        tracing::debug!("periodic update started");
-        let now = Instant::now();
+impl PoolFetching for UniswapV3PoolFetcher {
+    async fn fetch(
+        &self,
+        token_pairs: &HashSet<TokenPair>,
+        at_block: Block,
+    ) -> Result<Vec<PoolInfo>> {
+        let block_number = match at_block {
+            Block::Recent => self
+                .events
+                .lock()
+                .await
+                .store()
+                .last_event_block()
+                .await
+                .unwrap_or_default(),
+            Block::Number(number) => number,
+        };
 
-        let mut outdated_entries = self
-            .cache
+        // sanity check if events are up to date
+        let last_handled_block = self
+            .events
             .lock()
-            .unwrap()
-            .iter()
-            .filter(|(_, cached)| now.saturating_duration_since(cached.updated_at) > self.max_age)
-            .map(|(pool_id, cached)| (*pool_id, cached.requested_at))
-            .collect::<Vec<_>>();
-        outdated_entries.sort_by_key(|entry| std::cmp::Reverse(entry.1));
-        tracing::debug!("outdated_entries {:?}", outdated_entries);
+            .await
+            .last_handled_block()
+            .unwrap_or_default()
+            .0;
+        if block_number > last_handled_block {
+            tracing::debug!(
+                "can't get liquidity for block {} since the last handled block is {}",
+                block_number,
+                last_handled_block
+            );
+            if let Err(err) = self.events.run_maintenance().await {
```

Comment:
Could you add a comment for why we're running maintenance in PoolFetching::fetch? This feels surprising.

Reply:
Yes, it seems odd, should be just calling `update_events` but those two are equivalent (at least at the moment). Anyway, this call is increasing our chances of having events from the latest block indexed. (Because it could happen that we call `fetch` right after the new block is mined but `current_block_stream` did not detect it yet).  
Added comment.